LLEXT: more advanced detached section support

lyakh opened the pull request:
With this PR I have been able to move complex functions in extensions into detached sections with all the cross-referencing still working. I'll push a sample SOF PR, using this feature, later.

lyakh committed:
LLEXT: (cosmetic) reduce the scope of a variable

Move a variable calculation inside the block, where it is actually
used.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
LLEXT: Xtensa: fix logging level

Use the same logging level as in the rest of LLEXT.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
LLEXT: fix detached section cache synchronisation

Detached sections are used in situ without being copied to or
referenced from ext->mem[] LLEXT region arrays. Their caches must be
synchronised accordingly.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
LLEXT: fix advanced uses of detached sections

When detached sections are used, STB_GLOBAL relocations also have to
be processed depending on the relocation type. This commit unified
STB_GLOBAL and STB_LOCAL flows by making them use the same relocation
type parser, adds support for R_XTENSA_GLOB_DAT and R_XTENSA_JMP_SLOT
type relocations on Xtensa and fixes STT_SECTION address calculation
for such sections.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
LLEXT: Xtensa: add a Kconfig option to enable -fPIC

Currently when building LLEXT for Xtensa we use the -fPIC compiler
option, but this cannot be used when using detached sections in
extensions. Add a Kconfig option to switch between the two
compilation modes and switch -fPIC off when building relocatable
(partially linked) ELF binaries.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
LLEXT: preserve section headers

Move cached section headers to struct llext from struct llext_loader
to preserve them after llext_load() returns.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

teburd commented on subsys/llext/llext_link.c:
```diff
@@ -465,6 +465,20 @@ int llext_link(struct llext_loader *ldr, struct llext *ext, const struct llext_l
+		for (i = 0; i < ldr->sect_cnt; ++i) {
+			elf_shdr_t *shdr = ldr->sect_hdrs + i;
+
+			if (ldr_parm->section_detached(shdr)) {
```
Should this be wrapped in a compile time branch? E.g.
```c
if(IS_ENABLED(CONFIG_CACHE_MANAGEMENT) && ldr_param->section_detached) {
}
```

Otherwise we're doing more work than needed

lyakh replied:
> Should this be wrapped in a compile time branch? E.g.
> 
> ```c
> if(IS_ENABLED(CONFIG_CACHE_MANAGEMENT) && ldr_param->section_detached) {
> }
> ```
> 
> Otherwise we're doing more work than needed

@teburd there's a check for `ldr_parm->section_detached != NULL` above the loop, I don't think anyone will have that function supplied to LLEXT if they don't want to use it, but yes, we can add the check there too

teburd replied:
Ah the github diff didn't show that much, thanks!

pillo79 commented on cmake/compiler/xt-clang/target.cmake:
```diff
@@ -16,7 +14,20 @@ set(LLEXT_REMOVE_FLAGS
-  -fPIC
   -nostdlib
   -nodefaultlibs
 )
+
+if(CONFIG_LLEXT_BUILD_PIC)
+set(LLEXT_REMOVE_FLAGS ${LLEXT_REMOVE_FLAGS}
+  -fno-pic
+  -fno-pie
+)
+set(LLEXT_APPEND_FLAGS ${LLEXT_APPEND_FLAGS}
+  -fPIC
+)
+else()
+set(LLEXT_APPEND_FLAGS ${LLEXT_APPEND_FLAGS}
+  -ffreestanding
```
Interesting - never saw `-ffreestanding` before, so I have read about it... IIUC it makes the llext use the core's definition of the standard functions. Shouldn't this be the default for both PIC and non-PIC? (... and on other arches as well?)

Also, should we add `-fno-pic` here? Not sure what the Zephyr default is :man_shrugging:

[whatever the decision, also update `gcc/target_xtensa.cmake` to match]

pillo79 reviewed (commented):
Thanks, mostly LGTM and I love the `-fPIC` option implementation. A few comments below.

pillo79 reviewed (changes requested):
I said "mostly"... :innocent: but I disagree with the last commit in the series, and I really wish to find a better solution which works for everybody.

lyakh commented on subsys/llext/llext_link.c:
```diff
@@ -219,6 +226,15 @@ static void llext_link_plt(struct llext_loader *ldr, struct llext *ext, elf_shdr
 		}
 
+		/*
+		 * In some cases
+		 * text + got_offset
+		 * is the same as
+		 * (uintptr_t)llext_loaded_sect_ptr(ldr, ext, shdr->sh_info) + rel->r_offset
+		 * but not in all. Eventually we could figure out how to generalize the latter to
+		 * cover all cases and to possibly simplify the above calculation
```
@pillo79 hm, I did see them being different during development, but I re-added a check now and it never triggered... Looks like they are indeed equal, maybe I had a bug in that version still.

pillo79 replied:
If this is still a concern, I propose to replace this comment with a check, so we get useful debugging info for this issue.

lyakh replied:
@pillo79 how about "TODO: A should be the same as B, consider simplifying?"

lyakh commented on subsys/llext/llext_load.c:
```diff
@@ -737,8 +738,8 @@ int do_llext_load(struct llext_loader *ldr, struct llext *ext,
 	if (ldr->sect_hdrs_on_heap) {
 		llext_free(ldr->sect_hdrs);
+		ldr->sect_hdrs = NULL;
 	}
-	ldr->sect_hdrs = NULL;
 
```
@pillo79 yep, agree, I don't like that either and I expected objections against it. In your first option - you mean `llext_unload()`, right? But currently it has no access to the loader. So we'd need to either modify the function's signature by adding it as the second argument, or to add a pointer to the loader to `struct llext`?

pillo79 replied:
The loader struct can be dropped as soon as `llext_load` completes, so I would not create such a constraint. I was instead thinking about fully moving the field from `struct llext_loader` to `struct llext`. That way it is unambiguous that it's a permanent feature. 

The pointer could be freed at the end of `llext_load` unless a load_param field like `keep_section_info` (or whatever) was set, in which case it would be kept until `llext_unload`. 
What do you think?

lyakh replied:
@pillo79 in principle that sounds good, would work fine for me, but with that header array you anyway then have to scan it, you probably don't have section index to jump directly to it. And we do already have an API function `llext_find_section()` that scans that array. Its usefulness at least for SOF is currently limited by it only returning an offset, not a complete header. Ideally I would modify that function to also return an entire section header in a user-supplied parameter, but it's an official API (LLEXT is still experimental). We could add a new function for that and potentially deprecate / migrate over time.
Also, if we do move the pointer to `struct llext` maybe just keep it until freeing without a parameter - it isn't that much memory

pillo79 replied:
Great idea! Something along the lines of:
```
int llext_get_section_info(void **start, size_t *length, struct elf_shdr **shdr)
```
where each parameter is optional? 
Or even better, I guess we can have a `struct llext_section_info` with all those fields. It will make it easier to extend in the future without breaking compatibility. 
WDYT?

lyakh replied:
@pillo79 I think that function needs a `name` parameter ;-) And `llext` of course. But `length` is already in the section header? As for `start` - you mean its address in RAM? But that's only available if `peek()` is implemented? In SOF we do
```
offset = llext_find_section(...);
addr = llext_peek(ldr, offset);
```
So maybe only `llext`, `name` and `shdr` as parameters?

teburd replied:
I like it

pillo79 replied:
Sorry, I had a different use case in mind - for _loaded_ sections. LGTM as well!

lyakh replied:
> I like it

@teburd hm, but I've just realised, that it cannot work (without large-scale changes to LLEXT). All access to section header tables in LLEXT is currently performed via the loader. IIUC originally the loader was assumed to be used only for initial linking / relocation, only while `llext_load()` is running. Then we've relaxed that limitation by allowing the user to call `llext_seek()`, `llext_peek()`, `llext_read()` and potentially `llext_find_section()` - all with the loader as the first argument. But still we free all the dynamically allocated loader assets when exiting `llext_load()`. Maybe we indeed should extend the loader life-cycle and make an explicit `loader_free()`?

teburd replied:
This gets a bit awkward at some point as we're allocating in llext_load() but then I guess we're needing information past the load call. What lifetime are you looking for in the loader? Why?

pillo79 replied:
> we've relaxed that limitation by allowing the user to call `llext_seek()`, `llext_peek()`, `llext_read()` 

@lyakh These are used by LLEXT for ELF loading and they are [explicitly undocumented](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/llext/loader.h#L118). TBH I don't think an end user should use them :slightly_smiling_face: 

> we free all the dynamically allocated loader assets when exiting `llext_load()`

That should be fine, if we move `sect_hdrs` to `struct llext` as discussed. This is definitely a big commit, but should be automated with a `sed` command :slightly_smiling_face:

lyakh committed:
LLEXT: add llext_section_offset() to replace llext_find_section()

Now that section header cache is persistent, it can be used for
finding sections. Add a new function, similar to llext_find_section()
to use it and deprecate llext_find_section().

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

teburd reviewed (dismissed):
Makes sense to me

pillo79 commented on subsys/llext/llext.c:
```diff
@@ -23,12 +23,32 @@ static sys_slist_t _llext_list = SYS_SLIST_STATIC_INIT(&_llext_list);
 static struct k_mutex llext_lock = Z_MUTEX_INITIALIZER(llext_lock);
 
-ssize_t llext_find_section(struct llext_loader *ldr, const char *search_name)
+ssize_t llext_section_offset(struct llext_loader *ldr, struct llext *ext, const char *search_name)
```
The `find_section` API was IMO terribly broken, since it does not tell you the size associated with the section; it is unclear to me how to safely use it. :man_shrugging: 
Can we take this opportunity to provide a replacement where you get the full section header information instead? SOF can just refer to the `sh_offset` field in the returned struct.

```suggestion
const elf_shdr_t *llext_get_section_header(struct llext_loader *ldr, struct llext *ext, const char *search_name)
```

pillo79 commented on tests/subsys/llext/simple/src/test_llext_simple.c:
```diff
@@ -415,7 +415,7 @@ ZTEST(llext, test_find_section)
 	zassert_ok(res, "load should succeed");
 
-	section_ofs = llext_find_section(loader, ".data");
+	section_ofs = llext_section_offset(loader, ext, ".data");
```
Since `find_section` is not deprecated, both should be tested here.

lyakh replied:
we plan to deprecate it - once we've updated SOF to not call it to avoid twister failures

pillo79 reviewed (approved):
LGTM!
